### DISCUSSION TITLE
Storing scanner state within the tree nodes

### DISCUSSION THREAD

**Discussion by leonerd** [tree-sitter/tree-sitter#2054] [Q&A] [answered]
I have a question for which I *suspect* the answer is "it's impossible", and the only way to solve it would involve adding new tree-sitter API. At which point I'll probably propose it and have a go at implementing it, but first here it is.

Before I start I want to clarify my understanding of the existing scanner serialize/deserialize functions, in that they store one flat set of state across the entire parser. This state doesn't nest; isn't aware of the nested nature of tree nodes. I've tried testing this, and it appears my understanding is correct. Once you finish tree nodes and ascend back up out of "finished" inner sequences, you don't get back to previous lexer state.

## Background

I have [a parser](https://github.com/tree-sitter-perl/tree-sitter-pod) for a text markup language called POD. Conceptually it's quite similar to things like Markdown, having notations for "bold" and "italics" and "fixed-with code" and so on. The markers use a capital letter followed by an open angle-bracket (`X<`) to start, and a close angle-bracket to finish (`>`). I parse these out as "interior_sequence" nodes, which contain a "content" node:

```
This is B<bold>, this I<italic> and this is C<fixed-width code>.
---
(plain_paragraph
  (content
    (interior_sequence (content))
    (interior_sequence (content))
    (interior_sequence (content))))
```

All fairly simple so far. Lexical scanning is done using a simple `scanner.c` which stops scanning for "(content)" when it finds a capital letter followed by a `<`, so that an "(interior_sequence)" becomes a sequence of (start-letter, "<", content, ">").

They can be nested, and because the opens look different to the closes this is nice and simple to parse:

```
This is B<I<bold and italic>>
---
(plain_paragraph
  (content
    (interior_sequence
      (content
        (interior_sequence (content))))))
```

There's one feature I haven't yet implemented. POD allows you to use multiple `<`s to surround an interior sequence, requiring the same number to close it again. Within the sequence, any fewer markers than that are just taken literally. It allows you to easily write code fragments which might include the `>` character:

```
To compare two numbers you can use relational operations such as C<< $x >= $y >>.
```

This is mentioned in an Issue on the tracker - https://github.com/tree-sitter-perl/tree-sitter-pod/issues/1

If it weren't for the nesting ability of these sequences, then a `scanner.c` would be simple to implement. The scanner state just has to remember the count of open-brackets in an `int` or somesuch, and then look for that number to close, ignoring any smaller number as "just normal content". But the nesting ability makes that much more complicated, because now I have to remember the bracket count for the one we're closing, not just the most recently opened one.

In a perfect world, if there were no syntax errors, then I could keep a stack in the scanner and remember to push/pop it as appropriate. But given the way that `tree-sitter`'s error recovery works, especially while the user is busy editing the file within a text editor, I suspect this would very quickly get out of sync with the real shape of the tree nodes. Plus it means I have to store scanner state in a stack in the first place, which feels like a waste of storage when the tree of parse nodes is just sitting right there.

## The Actual Question

I feel this would be much simpler to implement if the scanner could store some extra data into a tree node, and later query it back relative to where it is now. In my case, when I see an open sequence of `<` characters, I'd want to store the count of these in my opening node. Then later, when scanning for content or a close sequence, I'd want to be able to ask tree-sitter for the context, by asking to retrive the data I previously stored in an earlier sibling node of the one I am currently scanning for, while ignoring any subtrees nested *within* it.

Such an API  might be exposed as a couple of functions provided by tree-sitter itself, taking the current TSLexer as context (because the scanner doesn't have anything else to hand):

```c
// Associate a scanner data payload with the node being returned now
void tree_sitter_scanner_store_node_payload(TSLexer *lexer, char *payload, size_t len);

// Retrieve a scanner data payload from an earlier sibling node of the one that would now be returned
size_t tree_sitter_scanner_fetch_node_payload(TSLexer *lexer, int node_sym, char *payload);
```

I'd then use this API in the following way:

```c
  /* having just recognised an X<<... sequence */

  uint8_t bracket_count = ...;
  tree_sitter_scanner_store_node_payload(lexer, &bracket_count, 1);
  lexer->result_symbol = TOKEN_SEQUENCE_START;
  return true;
```

```c
  if(valid_symbols[TOKEN_SEQUENCE_END]) {
    uint8_t bracket_count;
    tree_sitter_scanner_fetch_node_payload(lexer, TOKEN_SEQUENCE_START, &bracket_count);

    /* we can now look for an ending sequence of the right number of '>'s */
  }
```

Without having at least some ability to store and retrieve that scanner state which is aware of the tree nodes, I don't think it's possible to solve this problem.

## Non-Solutions

For this particular situation, there is one other "obvious" way to solve it. That is, rather than having just one pair of start/end tokens defined by my scanner, I'd pick some "reasonable" number, say ten, and define ten entire pairs of them for one bracket, two brackets, etc... Then when one of these gets returned as the start node, the parser state itself keeps track of which ending bracket I'm looking for, and the scanner could use that knowledge to know how many close brackets it can safely ignore. Yes, it would work in this *particular* situation, but that's a horrible solution - why do we arbitrarily pick ten? It makes the parser state have to have ten copies of all the values.

Most importantly: I don't want to entertain that as a solution here because I'm using this problem as a "simpler to explain" proxy to a much harder problem that has a lot more state to it, that's going to come up elsewhere in another parser of mine. The API proposed above would easily solve that problem too, whereas doing it this other way would involve creating many many *thousands* of duplicate tree states - clearly not a good solution.

**Comment by maxbrunsfeld** [comment]
That is what the serialize and deserialize functions are for. A copy of the scanner’s state is stored in each external token node, and when picking up from a previous parse, the scanner’s state is restored.

**Reply by leonerd** [reply]
I don't believe that's true. I have tried testing it just now.

I don't yet have parser state for counting the `<`s, so as a proxy to test it I just store the sequence letter that is being parsed, and printing that letter whenever it sees a close `>`. When parsing `B<I<foo>>` it prints the letter `I` both times - for both closing brackets. For this to be correct I'd expect to see `B` the second time, yet it does not.

```
scanner.c DEBUG: token(TOKEN_START_PLAIN)
scanner.c DEBUG: Starting an interior sequence 'B'
scanner.c DEBUG: token(TOKEN_INTSEQ_LETTER)
scanner.c DEBUG: token(TOKEN_INTSEQ_START)
scanner.c DEBUG: Starting an interior sequence 'I'
scanner.c DEBUG: token(TOKEN_INTSEQ_LETTER)
scanner.c DEBUG: token(TOKEN_INTSEQ_START)
scanner.c DEBUG: token(TOKEN_CONTENT_PLAIN)
scanner.c DEBUG: Ended an interior sequence started by 'I'
scanner.c DEBUG: token(TOKEN_INTSEQ_END)
scanner.c DEBUG: Ended an interior sequence started by 'I'
```

I would want this final line to say "Ended an interior sequence started by 'B'".

**Reply by leonerd** [reply]
I'm aware of those functions, as mentioned - but they don't act in a way that's aware of the nested structure of the resulting tree:

> Before I start I want to clarify my understanding of the existing scanner serialize/deserialize functions, in that they store one flat set of state across the entire parser. This state doesn't nest; isn't aware of the nested nature of tree nodes. I've tried testing this, and it appears my understanding is correct. Once you finish tree nodes and ascend back up out of "finished" inner sequences, you don't get back to previous lexer state.

**Reply by maxbrunsfeld** [reply, marked as answer]
It sounds like you need to store a stack in your external scanner. That works fine, and we do it for a number of existing languages, like Python (which stores a stack of indent lengths) and Ruby (which can store a stack of string delimiters).

**Comment by ahelwer** [comment]
To clarify - you are trying to improve the error handling of the external scanner, not having trouble parsing valid syntax?

**Reply by leonerd** [reply]
Mmm, well both. I'm trying to implement scanning of valid syntax. *A* way to implement it could be for me to create my own stack state in my scanner, but that has depth limit problems, and problems keeping in sync with what tree-sitter itself is doing. It's very limited and fragile. If I could get tree-sitter to store the state in nodes for me, then both these problems would be avoided.

**Reply by ahelwer** [reply]
Unfortunately the language you're trying to build is not context-free. You can see this because to write this rule in a BNF-like form you'd have to parameterize the rule:
```
marked_area(n) ::= <{n}, text(n-1), >{n};
```
or else simply define a finite number of the rules, as you mentioned:
```
marked_area_1 ::= <, text_0, >;
marked_area_2 ::= <<, text_1, >>;
...
```
Parameterized CFG rules require a TM to parse. Thus storing the state as a stack in your scanner is the way to go. I don't think you'll encounter any depth limits and it isn't particularly fragile. I personally just immediately return false if tree-sitter calls my external scanner during error correction and it works pretty well.

Similar annotation languages like markdown require users to escape special symbols such as `*` when they write them, like `\*`. Part of the skill of language design is keeping an eye toward ease of parsing. This is somewhat similar to this issue which was recently posted, also regarding a custom note-taking language which has ambiguity issues (even involving the same character, the < symbol!): https://github.com/tree-sitter/tree-sitter/issues/2047

**Reply by leonerd** [reply]
> Unfortunately the language you're trying to build is not context-free.

Indeed. That's what I was alluding to in the very beginning when I said "I think this is impossible currently, but ...".

> Parameterized CFG rules require a TM to parse. Thus storing the state as a stack in your scanner is the way to go. I don't think you'll encounter any depth limits and it isn't particularly fragile. I personally just immediately return false if tree-sitter calls my external scanner during error correction and it works pretty well.

I'm still primarily worried about whether the state in my stack will become out of sync with the tree that tree-sitter constructs. I know it has the ability to guess around the extra presence or unexpected absence of certain tree nodes at certain times, to make a best-effort guess at parsing the rest. I'm primarily worried that it will perform some error correction around bad syntax in a way that causes my independently-managed stack to become out of sync with the shape of the tree. This would be entirely solved if I could store that extra data in the tree *itself*, using the API I proposed above. Such a sync problem could thus never arise.

> Similar annotation languages like markdown require users to escape special symbols such as `*` when they write them, like `\*`. Part of the skill of language design is keeping an eye toward ease of parsing.

There is, perhaps, something to be said for such language design. An alternative take on it is that these "sortof feels like human text" should be designed primarily for humans, because that's what they're supposed to be, even if it means we have to take slightly more complex paths in order to get computers to understand them.

In any case, I don't think a project like `tree-sitter` gets to complain too much, when its main goal is to allow people to write parsers for all sorts of languages. These languages and situations will continue to exist, so it would be useful it tree-sitter was powerful enough to handle them. I believe the API shape I suggested above is an easy way to allow folks to write scanners that can handle it.

**Reply by leonerd** [reply]
Actually there's another thought here. While the numbers are small it probably doesn't matter, but there's the principle that if the scanner has to maintain its own stack, then the total stored state becomes unnecessarily quadratic in size.

For my simple use-case explained in this discussion it is less important, but there's other situations I have in mind that make it much worse. Suppose at some point in the tree we're ten levels deep. That means every external node in the tree has to store **all** of those ten levels of stack state out of the scanner, even though most of them are basically constant. During the nodes where it gets "pushed", that state is growing longer, all the nodes in the tallest part of the tree have all the levels, then finally they start to shrink again. Overall we're storing O(n²) of bytes.

Whereas, given the API I outline where I can ask tree-sitter to store an additional payload in one particular tree-node of mine, it doesn't have to duplicate all those previous levels. Now, we only have to store O(n) of additional bytes into the tree, and we only have to copy state in/out once per conceptual nesting or denesting operation. The stored state becomes a lot smaller and more efficient to manage. It feels like a clear win to me.

**Reply by ahelwer** [reply]
> There is, perhaps, something to be said for such language design. An alternative take on it is that these "sortof feels like human text" should be designed primarily for humans, because that's what they're supposed to be, even if it means we have to take slightly more complex paths in order to get computers to understand them.

This is a tricky position to hold, because the human writing your language has to be able to understand how to write it. And if you can't easily write the rules about how it should be parsed, how can a human understand the rules about how it should be written?

> I'm primarily worried that it will perform some error correction around bad syntax in a way that causes my independently-managed stack to become out of sync with the shape of the tree.

I was also concerned about this before writing my external scanner but you'll be happy to learn this is basically not that big of a problem. Error recovery works pretty well, certainly well enough for syntax highlighting and simple live AST analysis. If you're concerned you can also just zero out your external scanner state when tree-sitter calls it in error recovery mode, but I haven't found that to be necessary.

**Reply by leonerd** [reply]
> This is a tricky position to hold, because the human writing your language has to be able to understand how to write it. And if you can't easily write the rules about how it should be parsed, how can a human understand the rules about how it should be written?

But that's just the thing. I can very easily write the rules for humans. "After a letter and some number of `<`s, you need the same number of `>`s to close it; anything shorter is ignored as literal text". I'm only having trouble expressing the rules for the computer in tree-sitter, because tree-sitter doesn't have a good place for me to store the data for "the same number as ...". My proposed API above gives such a place.

---

I'm not disagreeing with you here - you have confirmed my suspicion that the problem is indeed impossible to solve with the current API shape. That's fine. I am proposing that it would be useful to add the API outlined above (or some variant like it), so that better, more powerful scanners can be written, in order to parse languages such as this and others, for which currently there is no viable solution.

There's a whole new set of things that I could write tree-sitter parsers for that currently I cannot, which that API would allow me to. This would make tree-sitter more useful, more applicable in more situations. I might just be misinterpreting your responses, but I'm having trouble understanding why it feels like you are opposed to such an idea.

**Reply by ahelwer** [reply]
It isn't impossible to solve, though. External scanners can contain arbitrary logic so can parse absolutely anything.
